src/collections/decodeableMap.ts

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (c) 2023, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+/**
+ * This is an extension of the Map class that can match keys whether they are encoded or decoded.
+ * Decoding the key can solve some edge cases in component fullNames such as Layouts and Profiles.
+ * See: https://github.com/forcedotcom/cli/issues/1683
+ *
+ * Examples:
+ *
+ * Given a map with entries:
+ * ```javascript
+ * 'layout#Layout__Broker__c-v1.1 Broker Layout' : {...}
+ * 'layout#Layout__Broker__c-v9%2E2 Broker Layout' : {...}
+ * ```
+ *
+ * `decodeableMap.has('layout#Layout__Broker__c-v1%2E1 Broker Layout')` --> returns `true`
+ * `decodeableMap.has('layout#Layout__Broker__c-v9.2 Broker Layout')` --> returns `true`
+ */
+export class DecodeableMap<K extends string, V> extends Map<string, V> {
+  // Internal map of decoded keys to encoded keys.
+  // E.g., { 'foo-v1.3 Layout': 'foo-v1%2E3 Layout' }
+  // This is initialized in the `keysMap` getter.
+  private internalkeysMap!: Map<string, string>;
+
+  private get keysMap(): Map<string, string> {
```

Comment:
if you init at declaration, you also wouldn't have to have a getter for a private prop--you could just use the prop.